### PR TITLE
Refactor query forcing

### PR DESCRIPTION
**Pull request by cjgillot**
The control flow in those functions was very complex, with several layers of continuations.

I tried to simplify the implementation, while keeping essentially the same logic.
Now, all code paths go through `try_execute_query` for the actual query execution.
Communication with the `dep_graph` and the live caches are the only difference between query getting/ensuring/forcing.

### CODE REVIEW

**Comment by jyn514** [comment]
@cjgillot does this need a perf run?

**Comment by cjgillot** [comment]
Should be perf-neutral, but a perf run can't hurt.

**Comment by jyn514** [comment]
Let's hold off then, it will get a perf run when it's merged and whoever does perf triage will see it if it's significant.

**Comment by Mark-Simulacrum** [comment]
Queries are generally quite hot, and it won't hurt - @bors try @rust-timer queue

File: compiler/rustc_query_system/src/query/caches.rs
Comment (bjorn3) [inline compiler/rustc_query_system/src/query/caches.rs:105]: `&key` can be `key`, right? `key` is already a reference.

File: compiler/rustc_middle/src/ty/query/mod.rs
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/query/mod.rs:191]: Shouldn't this be kept?

File: compiler/rustc_middle/src/ty/query/on_disk_cache.rs
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/query/on_disk_cache.rs:374]: This changes behavior. Is this necessary?

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: compiler/rustc_middle/src/ty/query/mod.rs
Comment (cjgillot) [reply compiler/rustc_middle/src/ty/query/mod.rs:191]: The assert has been moved inside `foce_query`.

**Comment by JohnCSimon** [comment]
ping from triage
@bjorn3  - can you address the requested changes from bjorn3? if you already have, can you please mark the change request resolved and alert the reviewer? thank you.

**Comment by bjorn3** [comment]
@cjgillot is the PR author, not me. Also all my review comments have been resolved.

**Comment by jyn514** [comment]
ping @lcnr - are you a good reviewer for this? If not, can you suggest someone, maybe @wesleywiser?

**Comment by lcnr** [comment]
yeah, would prefer to hand this off :sweat_smile: this ended up slipping under my radar, sry

r? @wesleywiser feel free to reassign if you also can't review this rn

**Comment by wesleywiser** [comment]
Apologies, it's been a busy week for me and I haven't had time to look at this. It will probably be a few more days before I can take an in-depth look so if you don't want to wait that long, please feel free to re-assign; I don't want to hold this up 🙂 

 I do see some regressions in the performance data. Would it be feasible to try to narrow down which changes are causing that?

File: compiler/rustc_query_system/src/query/plumbing.rs
Comment (wesleywiser) [inline compiler/rustc_query_system/src/query/plumbing.rs:391]: There's a lot going on in 6f001c7c204a33c0b8907f98e8f5d1211bd7c9b7. Can you leave some comments about what the goal of the changes in this commit are?

File: compiler/rustc_query_system/src/query/plumbing.rs
Comment (wesleywiser) [inline compiler/rustc_query_system/src/query/plumbing.rs:384]: It might be worth (temporarily) reverting this commit. If possible, I would like to have some idea of where the performance regressions perf.rlo is showing are coming from. If we can narrow it down, then we can evaluate more clearly if the lost performance is worth the upsides.

File: compiler/rustc_query_system/src/dep_graph/graph.rs
Comment (wesleywiser) [inline compiler/rustc_query_system/src/dep_graph/graph.rs:244]: This might be another one to back out temporarily to see if it's causing the regression.

**Review by wesleywiser — COMMENTED** [review]
Did a first pass review. Sorry for the delay!

File: compiler/rustc_query_system/src/query/plumbing.rs
Comment (cjgillot) [reply compiler/rustc_query_system/src/query/plumbing.rs:391]: I added a few comments in the commit message, and split a few parts into other commits.

File: compiler/rustc_query_system/src/dep_graph/graph.rs
Comment (cjgillot) [reply compiler/rustc_query_system/src/dep_graph/graph.rs]: `index_to_node` is just an array access. It should be trivial to optimise.

**Comment by cjgillot** [comment]
> Apologies, it's been a busy week for me and I haven't had time to look at this. It will probably be a few more days before I can take an in-depth look so if you don't want to wait that long, please feel free to re-assign; I don't want to hold this up slightly_smiling_face
> 
> I do see some regressions in the performance data. Would it be feasible to try to narrow down which changes are causing that?

I did not manage to track down the regression.

**Comment by cjgillot** [comment]
I did not measure since Wesley's comments: @bors try @rust-timer queue

**Comment by Dylan-DPC** [comment]
@cjgillot can you resolve the conflicts?

**Comment by Dylan-DPC** [comment]
oops too late :smile:

**Comment by bjorn3** [comment]
Overall a huge improvement of up to 3.1% on full builds and 0.6% on bootstrap time. Modest improvements of up to 0.4% on incr-full with one outlier regression of 0.4% and a couple up to 0.2% regressions. incr-unchanged shows regressions up to 2.0%. incr-patched shows regressions up to 1.0%.

**Comment by crlf0710** [comment]
@cjgillot  Now that #85319 is merged, what's next steps on this?

File: compiler/rustc_query_system/src/dep_graph/graph.rs
Comment (wesleywiser) [inline compiler/rustc_query_system/src/dep_graph/graph.rs:268]: The `unwrap()` can't fail since we're guarding this function in the calling code with `self.is_fully_enabled()`. Could you leave a comment to that effect?

**Comment by wesleywiser** [comment]
@cjgillot Do you have more changes locally (is this still WIP)?

**Comment by cjgillot** [comment]
@wesleywiser This PR somewhat evolved into a experimental staging ground. You can unassign yourself. I will extract relevant parts into other PRs when I get satisfactory results.

**Comment by cjgillot** [comment]
Last perf run shows no runtime modification, and -8% on rustc_query_impl compile time.
I find the resulting code much simpler, -90 lines.
r? @wesleywiser

**Comment by Mark-Simulacrum** [comment]
Thanks! I looked over each commit, and while the code is definitely still not easily readable I think this is moving us in the right direction for sure, great work. Bootstrap time improvement is also nice to see.

I also think that the perf-regression label can be considered acceptable - the vast majority of benchmarks are showing slight improvements, with only a few of the smaller benchmarks showing regressions (helloworld, await-call-tree), and those are quite small too.

@rustbot label: +perf-regression-triaged

@bors r+

**Comment by Mark-Simulacrum** [comment]
@bors r+

(I think bors missed my earlier comment...)

**Comment by Mark-Simulacrum** [comment]
@rustbot label: +perf-regression-triaged

The results are as expected before merging, with the majority being slightly positive, and the negative results limited to coercions-debug (typically noisy, though this may be a real data point) and helloworld (small, not very representative of normal code). Generally speaking, it seems like the improvements in bootstrap time and code quality are sufficient to accept these small regressions (and overall mixed results, given the wide swath of small improvements).

**Pull request merged**
